### PR TITLE
Box types update

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## Explanation
We are slowly adding more typescript support to the Box component. This PR includes some more additions
- Improves types in the `box.d.ts` file
- Converts `box.stories.js` => `box.stories.tsx` to see linting issues if any occur which will help identify any type issues
- Converts `box.test.js` => `box.test.tsx`
- Adds descriptions to `box.d.ts` file
- Fixes some propTypes in other files that will need to be changed once the Box component is converted to typescript

### Before

https://user-images.githubusercontent.com/8112138/225186616-ec6dd35e-eb4a-43ec-8acd-eae20535cbdd.mov

### After

https://user-images.githubusercontent.com/8112138/225186710-6b56a481-a2f3-4746-b192-2189848d19b7.mov

## Manual Testing Steps
- Go to latest build of Storybook in this PR
- Search `Box` in search panel
- See updates story and docs

- Pull branch and check `box.stories.tsx` and `box.d.ts` 

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/avatar-base/avatar-base.js [hunk 1]
```diff
@@ -46,17 +46,17 @@ AvatarBase.propTypes = {
...
+  borderColor: PropTypes.oneOf(Object.values(BorderColor)),
   /**
    * The color of the text inside the AvatarBase
    * Defaults to TextColor.textDefault
    */
-  color: Box.propTypes.color,
+  color: PropTypes.oneOf(Object.values(TextColor)),
   /**
    * Additional classNames to be added to the AvatarToken
    */
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-base/avatar-base.js:62]: Screenshot 2023-03-14 at 7 13 32 PM

File: ui/components/component-library/avatar-favicon/avatar-favicon.js [hunk 2]
```diff
@@ -76,7 +76,7 @@ AvatarFavicon.propTypes = {
    * The border color of the AvatarFavicon
    * Defaults to Color.transparent
    */
-  borderColor: Box.propTypes.borderColor,
+  borderColor: PropTypes.oneOf(Object.values(BorderColor)),
   /**
    * Additional classNames to be added to the AvatarFavicon
    */
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-favicon/avatar-favicon.js:82]: Screenshot 2023-03-14 at 7 14 38 PM

File: ui/components/component-library/avatar-icon/avatar-icon.js [hunk 3]
```diff
@@ -70,12 +70,12 @@ AvatarIcon.propTypes = {
...
+  backgroundColor: PropTypes.oneOf(Object.values(BackgroundColor)),
   /**
    * The color of the text inside the AvatarIcon
    * Defaults to TextColor.primaryDefault
    */
-  color: Box.propTypes.color,
+  color: PropTypes.oneOf(Object.values(TextColor)),
   /**
    * Additional classNames to be added to the AvatarIcon
    */
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-icon/avatar-icon.js:81]: This follows our other conventions of these propTypes

File: ui/components/component-library/avatar-network/avatar-network.js [hunk 4]
```diff
@@ -102,17 +102,17 @@ AvatarNetwork.propTypes = {
    * The background color of the AvatarNetwork
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-network/avatar-network.js:102]: This follows our other conventions of these propTypes. No propType errors in console.

File: ui/components/component-library/avatar-token/avatar-token.js [hunk 5]
```diff
@@ -102,17 +102,17 @@ AvatarToken.propTypes = {
    * The background color of the AvatarToken
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-token/avatar-token.js:102]: This follows our other conventions of these propTypes. No propType errors in console.

File: ui/components/component-library/badge-wrapper/badge-wrapper.stories.tsx [hunk 6]
```diff
@@ -315,44 +316,44 @@ export const AnchorElementShape = () => (
...
           style={{ width: 16, height: 16 }}
         />
       }
     >
       <Box
-        backgroundColor={Color.infoDefault}
+        backgroundColor={BackgroundColor.infoDefault}
         style={{ width: 40, height: 80 }}
       />
     </BadgeWrapper>
```
Comment (georgewrmarshall) [inline ui/components/component-library/badge-wrapper/badge-wrapper.stories.tsx:359]: Fixing type issues

File: ui/components/component-library/badge-wrapper/badge-wrapper.types.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import PropTypes from 'prop-types';
 import Box from '../../ui/box';
-import type { BoxProps } from '../../ui/box/box.d';
+import type { BoxProps } from '../../ui/box/box';
```
Comment (georgewrmarshall) [inline ui/components/component-library/badge-wrapper/badge-wrapper.types.ts:3]: Revert this change
```suggestion
import type { BoxProps } from '../../ui/box/box.d';
```

File: ui/components/ui/box/README.mdx [hunk 8]
```diff
@@ -14,43 +14,42 @@ Box is a utility component that can be used for layout or as a base for other UI
...
+| borderWidth        | 0, 1, 2, 4, 6, 8, 9, 10, 12 or array of numbers [1, 2] for responsive props                                                                                                                                                       | -                  |
+| borderRadius       | BorderRadius values or array of BorderRadius values for responsive props from [../ui/helpers/constants/design-system.js](https://github.com/MetaMask/metamask-extension/blob/develop/ui/helpers/constants/design-system.js)       | -                  |
+| borderStyle        | BORDER_STYLE values or array of BORDER_STYLE values for responsive props from [../ui/helpers/constants/design-system.js](https://github.com/MetaMask/metamask-extension/blob/develop/ui/helpers/constants/design-system.js)       | -                  |
+| as                 | The polymorphic `as` prop allows you to change the root HTML element of the Box component. Defaults to 'div'                                                                                                                      | 'div'              |
```
Comment (georgewrmarshall) [inline ui/components/ui/box/README.mdx:53]: Fixes prop table

File: ui/components/ui/box/box.d.ts [hunk 9]
```diff
@@ -80,51 +72,272 @@ export type SizeNumber =
...
+  /**
+   * The text-color of the Box component.
+   * Use TextColor enum from '../../../helpers/constants/design-system';
+   * Accepts responsive props in the form of an array.
+   */
   color?: TextColor | IconColor | string; // TODO: remove string when someone smarter figures out the issue with the color prop
+  /**
+   * The ref of the Box component.
+   */
+  ref?: React.RefObject<HTMLElement>;
```
Comment (georgewrmarshall) [inline ui/components/ui/box/box.d.ts:340]: Adding ref type

File: ui/components/ui/box/box.d.ts [hunk 10]
```diff
@@ -80,51 +72,272 @@ export type SizeNumber =
...
+  BorderColor?,
+  BorderColor?,
+];
+
+export type BorderRadiusArray = [
+  BorderRadius,
+  BorderRadius?,
+  BorderRadius?,
+  BorderRadius?,
+];
```
Comment (georgewrmarshall) [inline ui/components/ui/box/box.d.ts:104]: Mostly updating to allow for array props

File: ui/components/ui/card/README.mdx [hunk 11]
```diff
@@ -27,16 +27,19 @@ The Card component has a set of default props that should meet most card use cas
...
+  <Card borderStyle={BorderStyle.dashed}>
 
 // To remove or change padding
 <Card padding={0} />
 // All padding related props of the Box component will work
 
 // To change the background color
-<Card backgroundColor={Color.backgroundAlternative} />
+<Card backgroundColor={BackgroundColor.backgroundAlternative} />
 ```
```
Comment (georgewrmarshall) [inline ui/components/ui/card/README.mdx:45]: Small docs fix on Card component

File: ui/components/ui/card/card.js [hunk 12]
```diff
@@ -1,57 +1,43 @@
...
-  padding: Box.propTypes.padding,
   /**
    * The background color of the card
    * Defaults to Color.backgroundDefault
    */
-  backgroundColor: Box.propTypes.backgroundColor,
+  backgroundColor: PropTypes.oneOf(Object.values(BackgroundColor)),
   /**
    * The Card component accepts all Box component props
    */
```
Comment (georgewrmarshall) [inline ui/components/ui/card/card.js:43]: Updating card component to follow normal conventions

### Before


https://user-images.githubusercontent.com/8112138/225188956-23bab4b1-1fdb-404e-8364-3dec3558f656.mov

### After


https://user-images.githubusercontent.com/8112138/225189003-d6b325a0-8a89-4e70-9b99-eae09cab6d52.mov

File: ui/helpers/constants/design-system.ts [hunk 13]
```diff
@@ -108,6 +108,7 @@ export enum TextColor {
   textDefault = 'text-default',
   textAlternative = 'text-alternative',
   textMuted = 'text-muted',
+  overlayInverse = 'overlay-inverse',
```
Comment (georgewrmarshall) [inline ui/helpers/constants/design-system.ts:111]: Adding missing TextColor

**Review by darkwing — DISMISSED** [review]

**Review by DDDDDanica — DISMISSED** [review]

**Comment by georgewrmarshall** [comment]
@DDDDDanica and @darkwing code coverage is failing because a storybook file could I please get a reapprove? I think it's fine to merge with code coverage failing. We should update config so `*.stories.*` are ignored

**Review by DDDDDanica — APPROVED** [review]

**Review by DDDDDanica — APPROVED** [review]

**Review by darkwing — APPROVED** [review]
